### PR TITLE
price feeds and refactoring

### PR DESCRIPTION
**Pull request by mateigraura**

### CODE REVIEW

File: liquidity_pool/src/liquidity.rs [hunk 1]
```diff
@@ -54,73 +61,99 @@ pub trait LiquidityModule:
...
 
-        let nonce = self
-            .blockchain()
-            .get_current_esdt_nft_nonce(&self.blockchain().get_sc_address(), &borrows_token);
+        self.borrowed_amount()
+            .update(|total| *total += &total_borrowable);
 
-        // send debt position tokens
+        self.reserves(&pool_token_id)
+            .update(|total| *total -= &borrowable_amount);
```
Comment (catalinnnn) [inline liquidity_pool/src/liquidity.rs:107]: `borrowable_amount` is the value in dollars, hope I'm not wrong. Shouldn't it be `total_borrowable` here also? aka the amount of tokens? In whichever case, a better naming will definitely help here, would name them something like: `borrow_amount_in_dollars` and `borrow_amount_in_tokens`

File: liquidity_pool/src/tokens.rs [hunk 2]
```diff
@@ -137,48 +150,18 @@ pub trait TokensModule:
...
-            &self.lend_token().get(),
+    fn mint_position_tokens(&self, token_id: &TokenIdentifier, amount: &Self::BigUint) -> u64 {
+        self.send().esdt_nft_create(
+            token_id,
             amount,
             &BoxedBytes::empty(),
             &Self::BigUint::zero(),
             &BoxedBytes::empty(),
-            metadata,
+            &(),
```
Comment (catalinnnn) [inline liquidity_pool/src/tokens.rs:160]: Didn't know one could do this. Nice

File: lending_pool/src/lib.rs [hunk 3]
```diff
@@ -28,19 +29,17 @@ pub trait LendingPool:
...
-            .into_option()
-            .unwrap_or_else(|| self.blockchain().get_caller());
+        let initial_caller = self.caller_from_option_or_sender(caller);
 
-        require!(amount > 0, "amount must be greater than 0");
-        require!(!initial_caller.is_zero(), "invalid address provided");
+        self.require_amount_greater_than_zero(&amount)?;
+        self.require_non_zero_address(&initial_caller)?;
 
         let pool_address = self.get_pool_address(&asset);
```
Comment (dorin-iancu) [inline lending_pool/src/lib.rs:41]: Could use the already existing function `get_pool_address_non_zero`.

File: lending_pool/src/lib.rs [hunk 4]
```diff
@@ -75,16 +72,62 @@ pub trait LendingPool:
...
+
+        let borrow_token_pool_address = self.get_pool_address_non_zero(&asset_to_borrow)?;
+        let lend_token_pool_address = self.get_pool_address_non_zero(&payment_lend_id)?;
+
+        let metadata = self
+            .liquidity_pool_proxy(lend_token_pool_address.clone())
+            .get_interest_metadata(payment_nonce)
+            .execute_on_dest_context();
+
+        let ltv = self.asset_ltv(&collateral_token_id).get();
```
Comment (dorin-iancu) [inline lending_pool/src/lib.rs:99]: Maybe check if this is zero and return an error if so. That way, we don't even make the call if the mapping doesn't exist. (which would yield wrong results anyway)

File: lending_pool/src/lib.rs [hunk 5]
```diff
@@ -129,17 +172,15 @@ pub trait LendingPool:
     fn liquidate_endpoint(
         &self,
         liquidate_unique_id: BoxedBytes,
-        #[var_args] initial_caller: OptionalArg<Address>,
+        #[var_args] caller: OptionalArg<Address>,
         #[payment_token] asset: TokenIdentifier,
         #[payment_amount] amount: Self::BigUint,
     ) -> SCResult<()> {
```
Comment (dorin-iancu) [inline lending_pool/src/lib.rs:178]: This is more of a personal preference of mine, but I've seen it adopted in many places. The order of arguments should be:
- payment args
- "normal" arguments
- var args

File: lending_pool/src/router.rs [hunk 6]
```diff
@@ -142,6 +143,14 @@ pub trait RouterModule: proxy_common::ProxyCommonModule + factory::FactoryModule
         Ok(())
     }
 
+    #[only_owner]
+    #[endpoint(setAssetLTV)]
+    fn set_asset_ltv(&self, asset: TokenIdentifier, ltv: Self::BigUint) -> SCResult<()> {
```
Comment (dorin-iancu) [inline lending_pool/src/router.rs:148]: No need to return `SCResult<()>` here at all. I've done this mistake myself in the past.

File: liquidity_pool/src/library.rs [hunk 7]
```diff
@@ -1,6 +1,6 @@
 elrond_wasm::imports!();
 
-const BP: u32 = 1000000000;
+const BP: u32 = 1_000_000_000;
```
Comment (dorin-iancu) [inline liquidity_pool/src/library.rs:3]: :+1:

File: liquidity_pool/src/liquidity.rs [hunk 8]
```diff
@@ -54,73 +61,99 @@ pub trait LiquidityModule:
...
+
+        let pool_asset_dec_big = Self::BigUint::from(pool_asset_data.decimals as u64);
+        let total_borrowable = (&borrowable_amount / &pool_asset_data.price) / pool_asset_dec_big;
+        let asset_reserve = self.reserves(&pool_token_id).get();
 
-        let mut asset_reserve = self.reserves(&asset).get();
         require!(
-            asset_reserve != Self::BigUint::zero(),
-            "asset reserve is empty"
+            asset_reserve < total_borrowable,
```
[the remark this answers is not in the document]
Comment (mateigraura) [reply liquidity_pool/src/liquidity.rs]: good catch. done

File: liquidity_pool/src/liquidity.rs (hunk 1 above)
Comment (mateigraura) [reply liquidity_pool/src/liquidity.rs]: done

File: lending_pool/src/lib.rs (hunk 3 above)
Comment (mateigraura) [reply lending_pool/src/lib.rs]: done

File: lending_pool/src/lib.rs (hunk 4 above)
Comment (mateigraura) [reply lending_pool/src/lib.rs]: done. added `get_ltv_exists_and_non_zero`

File: lending_pool/src/lib.rs (hunk 5 above)
Comment (mateigraura) [reply lending_pool/src/lib.rs]: will be refactored entirely, but changed the args order.

File: lending_pool/src/router.rs (hunk 6 above)
Comment (mateigraura) [reply lending_pool/src/router.rs]: done

**Review by dorin-iancu — APPROVED** [review]

**Review by catalinnnn — APPROVED** [review]

**Pull request merged**